mlir/lib/Dialect/MIOpen/Transforms/AlignTiling.cpp

```diff
@@ -43,624 +45,433 @@
+LogicalResult MILARewritePattern::matchAndRewrite(linalg::GenericOp laGeneric,
+                                                  PatternRewriter &b) const {
+  Location loc = laGeneric.getLoc();
+
+  // 0. Test compatibility
+  // 0.0. Only fully parallel for now
+  for (StringRef iterType :
+       laGeneric.iterator_types().getAsValueRange<StringAttr>())
+    if (iterType != "parallel")
+      return failure();
+
+  Value out = *laGeneric.outputs().begin(); // may be another arg
+  // 0.1. Test compatibility,  Only 1 output supported
+  if (laGeneric.outputs().size() > 1)
+    return failure();
+
+  // 0.2. Sanity check, skip already fused.
+  for (auto inp : laGeneric.inputs()) {
+    if (auto fusedAlloc = inp.getDefiningOp<GpuAllocOp>()) {
+      LLVM_DEBUG(llvm::dbgs() << "Found existing fusion, bailing\n");
+      return failure();
     }
+  }
 
-    // 1. Trace input to threadwise_copy. Collect transforms (to be applied to
-    // other inputs).
-    // 1.1. Find the conv2d output
-    Value twinp;
-    bool v2 = true;
-    SmallVector<Value, 5> transforms;
-    for (auto pair : llvm::enumerate(laGeneric.inputs())) {
-      int32_t idx = pair.index();
-      Value inp = pair.value();
-      // 1.1. first trace to back to regs, then forward to twcopy
-      SmallVector<Value, 5> transforms_l;
-      Value twinp_t =
-          traceToThreadwiseCopy<miopen::ThreadwiseCopyV2Op>(inp, transforms_l);
-      if (!twinp_t) {
-        LLVM_DEBUG(llvm::dbgs() << "Didn't find threadwise copy v2\n");
-        transforms_l.clear();
-        twinp_t =
-            traceToThreadwiseCopy<miopen::ThreadwiseCopyOp>(inp, transforms_l);
-        if (twinp_t)
-          v2 = false;
+  SmallVector<AffineMap> idxMaps = laGeneric.getIndexingMaps();
+  // Output must be indexed by identity map for this to work
+  AffineMap outIdxMap = idxMaps.back();
+  if (!outIdxMap.isIdentity())
+    return failure();
+
+  // 1. Trace input to threadwise_copy. Collect transforms (to be applied to
+  // other inputs).
+  // 1.1. Find the conv2d output
+  ThreadwiseCopyV2Op copyOp;
+  for (auto pair : llvm::zip(idxMaps, laGeneric.inputs())) {
+    AffineMap inpIdxMap = std::get<0>(pair);
+    Value inp = std::get<1>(pair);
+    ThreadwiseCopyV2Op maybeCopy = traceToThreadwiseCopy(inp);
+    if (maybeCopy) {
+      if (copyOp) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Multiple generic inputs come from writeback\n");
+        return failure();
       }
-      if (twinp_t) {
-        // 1.2. Only one input should trace to twcopy
-        assert(!twinp);
-        twinp = twinp_t;
-        transforms = transforms_l;
-        break;
-      } else {
-        // 2.1. Test aligned input with output type
-        auto inpIdxMap = idxMaps[idx].template cast<AffineMapAttr>();
-        if (!checkCompatibleTypes(inp.getType(), inpIdxMap, out.getType(),
-                                  outIdxMap)) {
-          return fail;
-        }
+      if (!inpIdxMap.isIdentity()) {
```

Comment:
@jungpark-mlir can you confirm this case with MIXR (NCHW->NHWC).

Reply:
I could've sworn there's special handling for that way further up the pipe - but yeah, good call